src/hyp_map.rs

```diff
@@ -254,6 +257,72 @@ impl UmodeElfRegion {
     }
 }
 
+/// The U-mode Input Region is a private region of the page table used by the hypervisor to pass
+/// data to the current U-mode operation. This avoids mapping hypervisor data directly in
+/// U-mode. This area is written by the hypervisor and mapped read-only in U-mode.
+struct UmodeInputRegion {
+    /// Volatile Slice used to write to this area from the hypervisor.
+    vslice: VolatileSlice<'static>,
+}
+
+impl UmodeInputRegion {
+    /// Allocate hypervisor pages and map them read-only in U-mode VA space.
+    fn map(sv48: &FirstStagePageTable<Sv48>, hyp_mem: &mut HypPageAlloc) -> Self {
+        // Allocate pages.
+        let num_pages = PageSize::num_4k_pages(UMODE_INPUT_SIZE);
+        let pages = hyp_mem.take_pages_for_hyp_state(num_pages as usize);
+        let start = pages.base();
+
+        // Map the allocated pages as read-only in U-mode at UMODE_INPUT_START.
+        // UMODE_INPUT_START is 4k aligned (enforced by static assertion), round_down wil be a no-op.
+        let vaddr = PageAddr::with_round_down(
+            RawAddr::supervisor_virt(UMODE_INPUT_START),
+            PageSize::Size4k,
+        );
+        let pte_fields = PteFieldBits::leaf_with_perms(PteLeafPerms::UR);
+        // Unwrap okay: this is called once when we are creating the page-table. Guaranteed to be
+        // unmapped.
+        let mapper = sv48
+            .map_range(vaddr, PageSize::Size4k, num_pages, &mut || {
+                hyp_mem.take_pages_for_hyp_state(1).into_iter().next()
+            })
+            .unwrap();
+        for (virt, phys) in vaddr
+            .iter_from()
+            .zip(pages.base().iter_from())
+            .take(num_pages as usize)
+        {
+            // Safety: These pages are mapped read-only in the VA area reserved for the U-mode
+            // input region mappings. Hypervisor will write to these pages using the physical
+            // mappings and U-mode will read them through this mapping. Safe because these are
+            // per-CPU mappings, and when the hypervisor will be writing to these pages via the
+            // physical mappings no CPU will be able to access these pages through the U-mode
+            // mappings.
+            unsafe {
+                mapper.map_addr(virt, phys, pte_fields).unwrap();
+            }
+        }
+        // Safety: the range `(start..max_addr)` is mapped in U-mode as read-only and was uniquely
+        // claimed for this area from the hypervisor map above.
+        let vslice = unsafe {
+            VolatileSlice::from_raw_parts(start.raw().bits() as *mut u8, UMODE_INPUT_SIZE as usize)
+        };
+        Self { vslice }
+    }
+
+    // Writes `data` in the current CPU's U-mode Input Region.
+    fn store<T: DataInit>(&mut self, data: T) -> Result<(), Error> {
+        let vref = self.vslice.get_ref(0).map_err(Error::UmodeInput)?;
+        vref.store(data);
```

Comment:
nit: Is it a good idea to zero the area before writing in case the new data is smaller than the old?

Reply:
The guarantee that this is cleared is due on u-mode execute request. I understand that this might not sound straightforward, but this piece of code (as you correctly noted, this is just a move) was a complex interaction between three systems, so I preferred to have the mechanism here and the checking elsewhere.

In case you are going to go with a type that destroy on drop, that was one of my iterations during review. :-)